profiles/killbill/src/test/java/org/killbill/billing/jaxrs/TestEntitlement.java

```diff
@@ -1417,6 +1420,268 @@ public void testSubscriptionEventsWithCatalogChange() throws Exception {
+
+        // Retrieves with GET
+        Subscription subscription = subscriptionApi.getSubscription(entitlementJson.getSubscriptionId(), requestOptions);
+        Assert.assertEquals(subscription.getState(), EntitlementState.ACTIVE);
+        Assert.assertEquals(subscription.getEvents().size(), 4);
+        // The blocking state technically happens before the entitlement start
+        Assert.assertEquals(subscription.getEvents().get(0).getEventType(), SubscriptionEventType.SERVICE_STATE_CHANGE);
+        // Local date only for the subscription event
+        Assert.assertEquals(subscription.getEvents().get(0).getEffectiveDate(), new LocalDate(2012, 4, 25));
+        Assert.assertEquals(subscription.getEvents().get(1).getEventType(), SubscriptionEventType.START_ENTITLEMENT);
+        Assert.assertEquals(subscription.getEvents().get(1).getEffectiveDate(), new LocalDate(2012, 4, 25));
+        Assert.assertEquals(subscription.getEvents().get(2).getEventType(), SubscriptionEventType.START_BILLING);
+        Assert.assertEquals(subscription.getEvents().get(2).getEffectiveDate(), new LocalDate(2012, 4, 25));
+        Assert.assertEquals(subscription.getEvents().get(3).getEventType(), SubscriptionEventType.PHASE);
+        Assert.assertEquals(subscription.getEvents().get(3).getEffectiveDate(), new LocalDate(2012, 5, 25));
+
+        BlockingStates allSubscriptionBlockingStates = accountApi.getBlockingStates(accountJson.getAccountId(), ImmutableList.<BlockingStateType>of(BlockingStateType.SUBSCRIPTION),
+                                                                                    ImmutableList.<String>of(activateSubscriptionBS.getService()),
+                                                                                    AuditLevel.NONE,
+                                                                                    requestOptions);
+        Assert.assertEquals(allSubscriptionBlockingStates.size(), 1);
+        Assert.assertEquals(allSubscriptionBlockingStates.get(0).getStateName(), activateSubscriptionBS.getStateName());
+        Assert.assertFalse(allSubscriptionBlockingStates.get(0).isBlockEntitlement());
+        // DateTime for the blocking state
+        Assert.assertEquals(allSubscriptionBlockingStates.get(0).getEffectiveDate().compareTo(activateSubscriptionBS.getEffectiveDate()), 0);
+
+        // MOVE AFTER TRIAL
+        callbackServlet.pushExpectedEvents(ExtBusEventType.SUBSCRIPTION_PHASE,
+                                           ExtBusEventType.INVOICE_CREATION,
+                                           ExtBusEventType.INVOICE_PAYMENT_SUCCESS,
+                                           ExtBusEventType.PAYMENT_SUCCESS);
+        clock.addDays(30);
+        callbackServlet.assertListenerStatus();
+
+        // Cancel IMM (Billing EOT)
+        callbackServlet.pushExpectedEvents(ExtBusEventType.SUBSCRIPTION_CANCEL,
+                                           ExtBusEventType.ENTITLEMENT_CANCEL);
```

Comment:
I suppose the `ENTITLEMENT_CANCEL` is for `ActionType.REQUESTED` as the billing cancelation is EOT?

Reply:
Actually, the entitlement is effective right away (while indeed, billing cancellation is EOT).

Reply:
I meant: `ExtBusEventType.SUBSCRIPTION_CANCEL` is for `ActionType.REQUESTED` as the billing cancelation is EOT?

Reply:
Correct. It's in the JSON metadata of the event:

```
{"actionType":"REQUESTED","bundleExternalKey":"719cddd2-9774-4c87-8621-8aa526c562e7"}
```